New tutorial on parallelization and SIMWE (#76)
OSGeo/grass-tutorials: content/tutorials/parallelization/SIMWE_parallelization.qmd

    transport_coeff=f"transport_capacity_{huc12}",
    shear_stress=f"shear_stress_{huc12}",
    niterations=simulation_time,
    erosion_deposition=f"erdep_tmp_{huc12}",
    nwalkers=5 * gs.region()["cells"],
)
tools.r_grow(input=f"erdep_tmp_{huc12}", output=f"erdep_{huc12}", radius=-2.01)

basin_map = gj.Map()
basin_map.d_rast(map=f"erdep_{huc12}")
basin_map.d_legend(
    raster=f"erdep_{huc12}",
    flags="t",
    at=[10, 15, 40, 95],
    range=[-0.5, 0.5],
    digits=2,
    title="Erosion/dep. [kg/m2s]",
    fontsize=12,
)
basin_map.show()
```

![](SIMWE_images/basin_030401010402_erdep.webp)

Calculate total erosion for the subwatershed.
Negative values from [r.sim.sediment](https://grass.osgeo.org/grass-devel/manuals/r.sim.sediment.html)
output are treated as erosion and converted to sediment mass.
Summary statistics are computed using [r.univar](https://grass.osgeo.org/grass-devel/manuals/r.univar.html).

```{python}
# Erosion: extract negative values and convert to positive mass [kg/m^2]
tools.r_mapcalc(
    f"erosion_{huc12} = if(erdep_{huc12} < 0, abs(erdep_{huc12}) * {simulation_time * 60}, null())"
)
erosion = tools.r_univar(map=f"erosion_{huc12}", format="json")
```

> ```text
>  'null_cells': 418835,

Reviewer: `...` or `{` missing.

```suggestion
```text
...
 'null_cells': 418835,
```

It might be also little confusing to show the a result like this which is actually pretty printed Python dict, but the code does not suggests printing and suggests JSON (which has slightly different syntax).

 'cells': 599907,
 'min': 3.4630666334578564e-08,
 'max': 154011.390625,
 'range': 154011.39062496537,
 'mean': 98.83270611141184,
 'mean_of_abs': 98.83270611141184,
 'stddev': 1045.8673705761544,
 'variance': 1093838.556835879,
 'coeff_var': 1058.2199068769523,
 'sum': 17895835.761005566}
```

Deactivate the active raster mask, restoring operations to apply across the full region.

```{python}
mask.deactivate()
```

Display water depth (`depth`) and erosion-deposition (`erdep`) rasters for the selected subwatershed.

```{python}
basin_map = gj.InteractiveMap()
basin_map.add_raster(f"depth_{huc12}")
basin_map.add_raster(f"erdep_{huc12}")
basin_map.show()
```

![](SIMWE_images/basin_030401010402_both.webp)

# Parallelized hydrologic and sediment modeling

To efficiently model erosion and deposition across a large number of HUC12 subwatersheds,
we will use the workflow we just ran and create a
script that uses Python's `multiprocessing` module to parallelize the workflow.
Each subwatershed is processed independently in its own environment, which allows computations to run concurrently without interference.

> Each subwatershed needs to set different mask and computational region, however normally, those settings are global, and so for different mask and region for each parallel process, we will use mask and region context managers.

Reviewer: ```suggestion
Each subwatershed needs to set a different mask and computational region. However, those settings are usually global for each mapset. So, to use different masks and regions for each parallel process, we will use the mask and region context managers.
```

* Masking is handled using `MaskManager`, a [context manager for setting and managing raster mask](https://grass.osgeo.org/grass-devel/manuals/libpython/grass.script.html#grass.script.MaskManager),
making it possible to have custom mask for the current process. This feature is available only since GRASS 8.5.

    ```python
    with gs.MaskManager(mask_name=f"basin_{huc12}"):
        # Run actual computation with active mask.

>         gs.run_command(...)

Reviewer: In this tutorial, calling tools.r_sim_water or tools.r_mapcalc is fitting. Generally, I don't know what would be good analog to general run_command(...).

```suggestion
        tools.r_sim_water(...)
```

    ```

* Computational region is handled using `RegionManager`, a [context manager for setting and managing computational region](https://grass.osgeo.org/grass-devel/manuals/libpython/grass.script.html#grass.script.RegionManager),
making it possible to have custom region for the current process. This feature is available only since GRASS 8.5.

    ```python
    with gs.RegionManager(vector=f"basin_{huc12}"):
        # Run actual computation in the specified region.
        gs.run_command(...)

>     ```

Reviewer: While the previous text talks about both, mask is technically below region in the hierarchy, i.e., first comes region and the mask refines that selection, so I would go with region followed by mask by a list like this one.

## Putting it all together

We will put all the pieces together into a script that runs each subwatershed in parallel.
The resulting statistics for each subwatershed are collected and stored in a JSON file for further analysis.

```{python}
%%writefile script.py
import os
import json
from multiprocessing import Pool, cpu_count
from tqdm import tqdm

import grass.script as gs
from grass.tools import Tools

def compute(huc12):
    simulation_time = 180
    # set overwrite to True to rerun the workflow
    tools = Tools(overwrite=True, quiet=True)
    tools.v_extract(
        input="river_basins",
        output=f"basin_{huc12}",
        where=f"huc12 == '{huc12}'",
        flags="t",
    )

>     # Set the computational region to match the basin

Reviewer: The matching raster cells might be forth repeating:

```suggestion
    # Set the computational region to match the basin
    # while using the NED raster cells.
```

or: ...while using the NED raster cell size and alignment.

Maybe worth emphasizing above as well.

    with gs.RegionManager(vector=f"basin_{huc12}", raster="ned"):
        tools.v_to_rast(
            input=f"basin_{huc12}",
            output=f"basin_{huc12}",
            use="val",
        )
        with gs.MaskManager(mask_name=f"basin_{huc12}"):
            # Run actual computation with active mask.
            tools.r_proj(
                project="nlcd",
                mapset="PERMANENT",
                input="nlcd",
                output=f"nlcd_{huc12}",

>             )

Reviewer: Will this use the mask in any way? Maybe just put it to the region section to set the expectations right.

                detachment_coeff=f"detachment_coef_{huc12}",
                transport_coeff=f"transport_capacity_{huc12}",
                shear_stress=f"shear_stress_{huc12}",
                niterations=simulation_time,
                erosion_deposition=f"erdep_tmp_{huc12}",
                nwalkers=5 * region["cells"],
            )
            tools.r_grow(
                input=f"erdep_tmp_{huc12}",
                output=f"erdep_{huc12}",
                radius=-2.01,
            )
            # Erosion: extract negative values and convert to positive mass [kg/m^2]
            tools.r_mapcalc(
                expression=f"erosion_{huc12} = if(erdep_{huc12} < 0, abs(erdep_{huc12}) * {simulation_time * 60}, 0)",
            )
            erosion = tools.r_univar(map=f"erosion_{huc12}", format="json", flags="e")
            return {
                "huc12": huc12,
                "erosion_mean": erosion["mean"],
                "erosion_median": erosion["median"],
                "erosion_total": erosion["sum"] * region["nsres"] * region["ewres"],
            }

if __name__ == "__main__":
    tools = Tools()
    # The entire workflow will run in parallel,

>     # so this limits the number of threads the tools can use to 1.

Reviewer: ```suggestion
    # so this limits the number of threads each individual tool can use to 1.
```

    tools.g_gisenv(set="NPROCS=1")
    basins = tools.v_db_select(format="json", map="river_basins")["records"]
    huc12s = [basin["huc12"] for basin in basins]

>     # set the number of processes:

Reviewer: ```suggestion
    # set the number of processes to be used for the computation in total
```

    with Pool(processes=cpu_count()) as pool:
        result = list(tqdm(pool.imap(compute, huc12s), total=len(huc12s)))
        with open("result.json", "w") as fp:
            json.dump(result, fp)
```

Now execute the script, it will take some time.

```{python}
%run script.py
```

Load the resulting JSON file into a pandas dataframe and normalize the results, so that we can easily compare the subwatersheds.

```{python}
import pandas as pd
import json

with open("result.json") as f:
    stats = json.load(f)

df = pd.DataFrame(stats)
df["normalized_erosion"] = df.erosion_mean / max(df.erosion_mean)
df
```

Load the subwatershed layers into geopandas for visualization. Join the dataframe with the simulation values using the _huc12_ key and keep only subwatersheds that have computed erosion values.

```{python}
import geopandas as gpd

gdf = gpd.read_file(hydro_filename, layer="WBDHU12")
gdf = gdf.merge(df, on="huc12", how="left")
# JSON does not support Timestamp type

> gdf = gdf.drop(columns=["loaddate"])

Reviewer: While not relevant to the tutorial, this seems to need just little more info.

> gdf = gdf[gdf["erosion_total"].notna()]

Reviewer: Just comment what this is doing. Remove results without erosion? But why are these here anyway?

```

Let's visualize the data with folium:

```{python}
import folium
import branca.colormap as cm

metric = "normalized_erosion"
color_scale = cm.linear.OrRd_09.scale(min(gdf[metric]), max(gdf[metric]))
color_scale.caption = "Min-max normalized mean erosion"
# get region center as latitude, longitude
region = tools.g_region(vector="river_basins", format="json", flags="b")
m = folium.Map(location=[region["ll_clat"], region["ll_clon"]], zoom_start=9)

tooltip = folium.GeoJsonTooltip(
    fields=["huc12", metric],
    aliases=["HUC12", "Normalized mean erosion:"],
    localize=True,
    sticky=False,
    labels=True,
    style="""
        background-color: #F0EFEF;
        border: 2px solid black;
        border-radius: 3px;
        box-shadow: 3px;
    """,
    max_width=800,
)

g = folium.GeoJson(
    gdf,
    style_function=lambda x: {
        "fillColor": color_scale(x["properties"][metric]),
        "color": "black", "weight": 1,
        "fillOpacity": 0.5,
    },
    tooltip=tooltip,
).add_to(m)
color_scale.add_to(m)
folium.LayerControl().add_to(m)
m
```

![](SIMWE_images/webmap.webp)

We can now identify subwatersheds that have larger erosion values and are potentially

> the source of the source of increased sediment loads downstream, contributing to water quality degradation and habitat disruption.

Reviewer: Say this upfront even with a caveat that this will be a somewhat limited analysis.

* Using mask to mask out subwatershed area
* Parallelization of a computational workflow
* Overland water flow and sediment transport simulation

::: {.callout-note title="How to run this tutorial?"}
This tutorial is prepared to be run in a Jupyter notebook locally
or using services such as Google Colab. You can [download the notebook here](SIMWE_parallelization.ipynb).

If you are not sure how to get started with GRASS, checkout the tutorials [Get started with GRASS & Python in Jupyter Notebooks](../get_started/fast_track_grass_and_python.qmd)
and [Get started with GRASS in Google Colab](../get_started/grass_gis_in_google_colab.qmd).
:::

# Setup

Start with importing Python packages.
To import the _grass_ package, you need to tell Python where the GRASS Python package is
(can be skipped for some environment).

```{python}
# import standard Python packages
import os
import sys
import subprocess

# check where GRASS Python packages are and add them to path
sys.path.append(
    subprocess.check_output(["grass", "--config", "python_path"], text=True).strip()
)
# import GRASS Python packages
import grass.script as gs
import grass.jupyter as gj
from grass.tools import Tools
```

> We will switch our working directory to a directory with enough space, this is where we will download the datsets.

Author: ```suggestion
We will switch our working directory to a directory with enough space, this is where we will download the datasets.
```

color_scale.caption = "Min-max normalized mean erosion"
# get region center as latitude, longitude
region = tools.g_region(vector="river_basins", format="json", flags="b")
m = folium.Map(location=[region["ll_clat"], region["ll_clon"]], zoom_start=9)

tooltip = folium.GeoJsonTooltip(
    fields=["huc12", metric],
    aliases=["HUC12", "Normalized mean erosion:"],
    localize=True,
    sticky=False,
    labels=True,
    style="""
        background-color: #F0EFEF;
        border: 2px solid black;
        border-radius: 3px;
        box-shadow: 3px;
    """,
    max_width=800,
)

g = folium.GeoJson(
    gdf,
    style_function=lambda x: {
        "fillColor": color_scale(x["properties"][metric]),
        "color": "black", "weight": 1,
        "fillOpacity": 0.5,
    },
    tooltip=tooltip,
).add_to(m)
color_scale.add_to(m)
folium.LayerControl().add_to(m)
m
```

![](SIMWE_images/webmap.webp)

We can now identify subwatersheds that have larger erosion values and are potentially
the source of increased sediment loads downstream, contributing to water quality degradation and habitat disruption.

## Acknowledgements

> The development of this tutorial was supported bym NSF Award #2322073, granted to Natrx, Inc. and USDA NRCS award NR233A750023C043.

Author: ```suggestion
The development of this tutorial was supported by NSF Award #2322073, granted to Natrx, Inc. and USDA NRCS award NR233A750023C043.
```
